### PR TITLE
Elizabeth/scale decode

### PR DESCRIPTION
**Pull request by noot**
## Changes
Implemented SCALE decoding for compact integers < 2**64 as well as byte arrays and booleans. 

## Tests:
```
go test ./codec/...
```

### CODE REVIEW

File: codec/decode.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		}
+	} else if mode == 3 { // encoding of length: big-integer mode
+		length, err := DecodeInteger(b)
+
+		if err == nil {
+			// get the length of the encoded length
+			topSixBits := (binary.LittleEndian.Uint16(b) & 0xff) >> 2
+			byteLen := topSixBits + 4
+
+			if (length < 1 << 30 || int64(len(b)) < length + int64(byteLen)) {
```
Comment (dutterbutter) [inline codec/decode.go]: You could remove the parenthesis `if length < 1 << 30 || int64(len(b)) < length + int64(byteLen)`

File: codec/decode.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		o = int64(binary.LittleEndian.Uint32(b[1 : byteLen+1]))
+	} else if byteLen > 4 && byteLen < 8 {
+		upperBytes := make([]byte, 8-byteLen)
+		upperBytes = append(b[5:byteLen+1], upperBytes...)
+		upper := int64(binary.LittleEndian.Uint32(upperBytes)) << 32
+		lower := int64(binary.LittleEndian.Uint32(b[1:5]))
+		o = upper + lower
+	}
+
+	if o == 0 {
```
Comment (ansermino) [inline codec/decode.go]: How do you know we didn't just decode `0`?

File: codec/decode.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// DecodeBool accepts a byte array representing a SCALE encoded bool and performs SCALE decoding
+// of the bool then returns it. if invalid, return false and an error
+func DecodeBool(b byte) (bool, error) {
+	if b == 1 {
+		return true, nil
+	} else if b == 0 {
+		return false, nil
+	}
```
Comment (ansermino) [inline codec/decode.go]: Perhaps?
```
else {
    return false, Error(...)
}
```
Input is not boolean so no gurantees

File: codec/decode_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	val    []byte
+	output []byte
+}
+
+type decodeBoolTest struct {
+	val    byte
+	output bool
+}
+
+var decodeIntTests = []decodeIntTest{
```
Comment (ansermino) [inline codec/decode_test.go]: Just for educational purposes:
```
... = []struct{
  val byte
  output bool
}{
  {val: []byte{0x00}, output: int64(0)}
  ...
}
```

File: codec/decode.go (hunk 2 above)
Comment (noot) [reply codec/decode.go]: if we decoded 0 it should have happened in the first case (1 byte mode), it is invalid if it happens in the last case. so I think it's ok to throw an error here, as the bytes in a large integer >2**30 shouldn't be zero

File: codec/decode.go (hunk 3 above)
Comment (noot) [reply codec/decode.go]: gometalinter doesn't like that :/ but yeah I figured if the input byte isn't 1 or 0 just return false and error

File: codec/decode_test.go (hunk 4 above)
Comment (noot) [reply codec/decode_test.go]: nice

File: .gitignore [hunk 5]
```diff
@@ -10,13 +10,16 @@
...
 .idea
 
-# VSCode
+# VS Code
 .vscode
 
-go-pre.db
\ No newline at end of file
+# Sublime
+*.sublime*
```
Comment (priom) [inline .gitignore]: why did u take off `go-pre.db`, we shouldn't commit the database

Comment (noot) [reply .gitignore]: I didn't, might have had an old version. I can add it back

**Pull request merged**